lib/signaling/v2/cancelableroomsignalingpromise.js

```diff
@@ -12,120 +12,133 @@ const {
 
 const { flatMap } = require('../../util');
 
-function createCancelableRoomSignalingPromise(token, wsServer, localParticipant, iceServerSource, encodingParameters, preferredCodecs, options) {
+function createCancelableRoomSignalingPromise(token, wsServer, localParticipant, encodingParameters, preferredCodecs, options) {
   options = Object.assign({
     PeerConnectionManager: DefaultPeerConnectionManager,
     RoomV2: DefaultRoomV2,
     Transport: DefaultTransport
   }, options);
 
-  let transport;
-
-  const PeerConnectionManager = options.PeerConnectionManager;
-  const RoomV2 = options.RoomV2;
-
-  const peerConnectionManager = new PeerConnectionManager(iceServerSource, encodingParameters, preferredCodecs, options);
-
+  const { PeerConnectionManager, RoomV2, Transport, iceServers } = options;
+  const peerConnectionManager = new PeerConnectionManager(encodingParameters, preferredCodecs, options);
   const trackSenders = flatMap(localParticipant.tracks, trackV2 => [trackV2.trackTransceiver]);
-
-  peerConnectionManager.setConfiguration(options);
   peerConnectionManager.setTrackSenders(trackSenders);
 
   const cancelationError = new Error('Canceled');
 
-  return new CancelablePromise(function onCreate(resolve, reject, isCanceled) {
-    peerConnectionManager.createAndOffer().then(function createAndOfferSucceeded() {
-      // NOTE(mmalavalli): PeerConnectionManager#createAndOffer() queues the
-      // initial offer in the event queue for the 'description' event. So,
-      // we are dequeueing to prevent the spurious 'update' message sent by
-      // the client after connecting to a room.
-      peerConnectionManager.dequeue('description');
+  let transport;
 
-      return new Promise((resolve, reject) => {
+  return new CancelablePromise((resolve, reject, isCanceled) => {
+    return new Promise((resolve, reject) => {
+      const onIced = iceServers => {
         if (isCanceled()) {
```

Comment:
@makarandp0 This is true when you are raising an exception in the `then()` success callback. But here, we already have access to the `reject` callback of the parent promise.

Reply:
@manjeshbhargav, I think you are shadowing outer `reject` with inner variable - We should really enable warning to not allow variable shadowing - but that means outer promise wont get rejected immediately. It will only get rejected in the `catch`  of the inner promise.

isn't 
```    reject(cancellationError);
          return Promise.reject(cancellationError);
``` 
same as simpler
``` 
throw cancellationError;
```
Thinking a bit about this, do we really need this inner `Promise` ?

Reply:
Another suggestion is to move onIced outside the CancellablePromise.
``` 
const peerConnectionManager = new PeerConnectionManager(encodingParameters, preferredCodecs, options);
const trackSenders = flatMap(localParticipant.tracks, trackV2 => [trackV2.trackTransceiver]);
peerConnectionManager.setTrackSenders(trackSenders);
let cancellationError = null; // this will get set in onCancel handler if caller cancels.
const onIced = iceServers => {
  log.debug('Got ICE servers:', iceServers);
  if (cancellationError) {
    throw cancellationError;
  }
  options.iceServers = iceServers;
  peerConnectionManager.setConfiguration(options);

  return peerConnectionManager.createAndOffer().then(() => {
    log.debug('createAndOffer() succeeded.');
    if (cancellationError) {
      throw cancellationError;
    }
    // NOTE(mmalavalli): PeerConnectionManager#createAndOffer() queues the
    // initial offer in the event queue for the 'description' event. So,
    // we are dequeueing to prevent the spurious 'update' message sent by
    // the client after connecting to a room.
    peerConnectionManager.dequeue('description');
  }).catch(error => {
    log.error('createAndOffer rejected', error);
    throw error;
  });
};

return new CancelablePromise((resolve, reject, isCanceled) => {
  ...
}, () => {
  if (transport) {
    transport.disconnect();
    cancellationError = new Error('Canceled');
  }
});
```

Reply:
@makarandp0 `onIced` has to be declared inside the CancelablePromise because, we need the `isCanceled` callback to reject the CancelablePromise. Also, we can't throw the first `cancellationError` because it is not being done within a Promise executor. Therefore we need `Promise.reject(cancellationError)`.